azurerm/data_source_hdinsight_cluster.go

```diff
@@ -0,0 +1,157 @@
+						"enabled": {
+							Type:     schema.TypeBool,
+							Computed: true,
+						},
+						"username": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"password": {
+							Type:      schema.TypeString,
+							Computed:  true,
+							Sensitive: true,
+						},
+					},
+				},
+			},
+
+			"tags": tagsForDataSourceSchema(),
+
+			"edge_ssh_endpoint": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"https_endpoint": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"ssh_endpoint": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceArmHDInsightClusterRead(d *schema.ResourceData, meta interface{}) error {
+	clustersClient := meta.(*ArmClient).hdinsightClustersClient
+	configurationsClient := meta.(*ArmClient).hdinsightConfigurationsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	resourceGroup := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+
+	resp, err := clustersClient.Get(ctx, resourceGroup, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			log.Printf("[DEBUG] HDInsight Cluster %q was not found in Resource Group %q - removing from state!", name, resourceGroup)
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error retrieving HDInsight Cluster %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	configuration, err := configurationsClient.Get(ctx, resourceGroup, name, "gateway")
+	if err != nil {
+		return fmt.Errorf("Error retrieving Configuration for HDInsight Cluster %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	d.SetId(*resp.ID)
+
+	d.Set("name", name)
+	d.Set("resource_group_name", resourceGroup)
+	if location := resp.Location; location != nil {
+		d.Set("location", azureRMNormalizeLocation(*location))
+	}
+
+	if props := resp.Properties; props != nil {
+		d.Set("cluster_version", props.ClusterVersion)
+		d.Set("tier", string(props.Tier))
+
+		if def := props.ClusterDefinition; def != nil {
+			d.Set("component_versions", flattenHDInsightsDataSourceComponentVersions(def.ComponentVersion))
+			if kind := def.Kind; kind != nil {
+				d.Set("kind", strings.ToLower(*kind))
+			}
+			if err := d.Set("gateway", azure.FlattenHDInsightsConfigurations(configuration.Value)); err != nil {
+				return fmt.Errorf("Error flattening `gateway`: %+v", err)
+			}
+		}
+
+		edgeNodeSshEndpoint := azure.FindHDInsightConnectivityEndpoint("EDGESSH", props.ConnectivityEndpoints)
+		d.Set("edge_ssh_endpoint", edgeNodeSshEndpoint)
+		httpEndpoint := azure.FindHDInsightConnectivityEndpoint("HTTPS", props.ConnectivityEndpoints)
+		d.Set("https_endpoint", httpEndpoint)
+		sshEndpoint := azure.FindHDInsightConnectivityEndpoint("SSH", props.ConnectivityEndpoints)
+		d.Set("ssh_endpoint", sshEndpoint)
+	}
+
+	flattenAndSetTags(d, resp.Tags)
+
+	return nil
+}
+
+func flattenHDInsightsDataSourceComponentVersions(input map[string]*string) map[string]string {
```

Comment:
I'm not going to do this for the moment because I'm planning on opening a GH issue with the bugs/suggested enhancements I've found when building against the API - of which this is one (this object could be strongly typed, since the keys appear to be hard-coded)